File: userfctR/user_DrivenJoints.m
function [q,qd,qdd] = user_DrivenJoints(mbs_data,tsim)
% --------------------------
% UCL-CEREM-MBS
%
% @version MBsysLab_m 1.7.a
%
% Creation : 2005
% Last update : 30/09/2008
% -------------------------
%
%[q,qd,qdd] = user_DrivenJoints(mbs_data,tsim)
%
% mbs_data : multibody data structure
% tsim : current time
%
% q, qd, qdd : updated column vectors of generalized coordinates
%
%
% mbs_data.q : generalized coordinates [column vector]
% mbs_data.qd : generalized velocities [column vector]
% mbs_data.qdd : generalized accelerations [column vector]
% mbs_data.nqc : number of driven variables
% mbs_data.qc : indices of driven variables [column vector]

global MBS_user MBS_info

q   = mbs_data.q;
qd  = mbs_data.qd;
qdd = mbs_data.qdd;

%/*-- Begin of user code --*/

T = 1; %[s]
omega = 2*pi/(2*T); %[rad/s]

id = mbs_get_joint_id(MBS_info,'R2_FR');
if(mod(tsim,2*T) < T)
        q(id)    = pi*(-cos(omega*tsim)+1+floor(tsim));
        qd(id)      = pi*omega*sin(omega*tsim);
        qdd(id)     = pi*omega*omega*cos(omega*tsim);
    else
        q(id)   = pi*(1+floor(tsim));
        qd(id)  = 0;
        qdd(id) = 0;
end


id = mbs_get_joint_id(MBS_info,'R2_FL');
if(mod(tsim,2*T) >= T)
        q(id)    = pi*(-cos(omega*tsim+pi)+floor(tsim));
        qd(id)      = pi*omega*sin(omega*tsim+pi);
        qdd(id)     = pi*omega*omega*cos(omega*tsim+pi);
    else
        q(id)   = pi*(floor(tsim));
        qd(id)  = 0;
        qdd(id) = 0;
end


id = mbs_get_joint_id(MBS_info,'R2_RR');
if(mod(tsim,2*T) >= T)
        q(id)    = pi*(-cos(omega*tsim+pi)+floor(tsim));
        qd(id)      = pi*omega*sin(omega*tsim+pi);
        qdd(id)     = pi*omega*omega*cos(omega*tsim+pi);
    else
        q(id)   = pi*(floor(tsim));
        qd(id)  = 0;
        qdd(id) = 0;
end


id = mbs_get_joint_id(MBS_info,'R2_RL');
if(mod(tsim,2*T) < T)
        q(id)       = pi*(-cos(omega*tsim)+1+floor(tsim));
        qd(id)      = pi*omega*sin(omega*tsim);
        qdd(id)     = pi*omega*omega*cos(omega*tsim);
    else
        q(id)   = pi*(1+floor(tsim));
        qd(id)  = 0;
        qdd(id) = 0;
end


% i = ...
%     q(i) = f(tsim) ...
%     qd(i) = f'(tsim) ...
%     qdd(i) = f''(tsim) ...

%/*-- End of user code --*/

return

File: Standalone/src/project/user_files/user_DrivenJoints.c
//---------------------------
// UCL-CEREM-MBS
//
// @version MBsysLab_s 1.7.a
//
// Creation : 2006
// Last update : 01/10/2008
//---------------------------

#include "simu_def.h"
//#include <math.h>

#define GO

// Sets the pos, vel, acc of driven joints (if different from 0)
void user_DrivenJoints(MBSdataStruct *MBSdata, double tsim)
{
    UserIOStruct *uvs;
    double T = 1.0; //[s]
	double My_PI = 3.1415926;
	double omega = 2*My_PI/(2*T); //[rad/s]
	int id = 0;

	uvs = MBSdata->user_IO;


#ifdef GO
	//Front Right Leg
//     id = R2_FR;
// 	if(fmod(tsim,2*T) < T)
// 	{
//         MBSdata->q[id]    	= My_PI*(-cos(omega*tsim)+1+floor(tsim));
//         MBSdata->qd[id]      = My_PI*omega*sin(omega*tsim);
//         MBSdata->qdd[id]     = My_PI*omega*omega*cos(omega*tsim);
//     }
//     else
//     {
//         MBSdata->q[id]   = My_PI*(1+floor(tsim));
//         MBSdata->qd[id]  = 0;
//         MBSdata->qdd[id] = 0;
//     }

    //Front Left Leg
 //    id = R2_FL;
 //    if(fmod(tsim,2*T) >= T)
	// {
 //        MBSdata->q[id]    	= My_PI*(-cos(omega*tsim + My_PI)+floor(tsim));
 //        MBSdata->qd[id]      = My_PI*omega*sin(omega*tsim + My_PI);
 //        MBSdata->qdd[id]     = My_PI*omega*omega*cos(omega*tsim + My_PI);
 //    }
 //    else
 //    {
 //        MBSdata->q[id]   = My_PI*(floor(tsim));
 //        MBSdata->qd[id]  = 0;
 //        MBSdata->qdd[id] = 0;
 //    }

    //Rear Right Leg
 //    id = R2_RR;
 //    if(fmod(tsim,2*T) >= T)
	// {
 //        MBSdata->q[id]    	 = My_PI*(-cos(omega*tsim + My_PI)+floor(tsim));
 //        MBSdata->qd[id]      = My_PI*omega*sin(omega*tsim + My_PI);
 //        MBSdata->qdd[id]     = My_PI*omega*omega*cos(omega*tsim + My_PI);
 //    }
 //    else
 //    {
 //        MBSdata->q[id]   = My_PI*(floor(tsim));
 //        MBSdata->qd[id]  = 0;
 //        MBSdata->qdd[id] = 0;
 //    }

    //Rear Left Leg
 //    id = R2_RL;
	// if(fmod(tsim,2*T) < T)
	// {
 //        MBSdata->q[id]    	= My_PI*(-cos(omega*tsim)+1+floor(tsim));
 //        MBSdata->qd[id]      = My_PI*omega*sin(omega*tsim);
 //        MBSdata->qdd[id]     = My_PI*omega*omega*cos(omega*tsim);
 //    }
 //    else
 //    {
 //        MBSdata->q[id]   = My_PI*(1+floor(tsim));
 //        MBSdata->qd[id]  = 0;
 //        MBSdata->qdd[id] = 0;
 //   }
#else
		//Front Right Leg
//    id = R2_FR;
//	    MBSdata->q[id]    	= 0;
//        MBSdata->qd[id]  = 0;
//        MBSdata->qdd[id] = 0;
//
//
//    //Front Left Leg
//    id = R2_FL;
//  	    MBSdata->q[id]    	= 0;
//        MBSdata->qd[id]  = 0;
//        MBSdata->qdd[id] = 0;
//
//    //Rear Right Leg
//    id = R2_RR;
// 	    MBSdata->q[id]    	= 0;
//        MBSdata->qd[id]  = 0;
//        MBSdata->qdd[id] = 0;
//
//    //Rear Left Leg
//    id = R2_RL;
//	    MBSdata->q[id]    	= 0;
//        MBSdata->qd[id]  = 0;
//        MBSdata->qdd[id] = 0;
//

#endif

}
